br refactor: translate snapshot br (#11392)
pingcap/docs: br/br-snapshot-guide.md

---
title: Snapshot Back up and Restore Guide
summary: Learn about how to back up and restore TiDB snapshots using the br command-line tool.
aliases: ['/tidb/dev/br-usage-backup/','/tidb/dev/br-usage-restore/','/tidb/dev/br-usage-restore-for-maintain/', '/tidb/dev/br-usage-backup-for-maintain/']
---

# Snapshot Back up and Restore Guide

This document describes how to back up and restore TiDB snapshots using the `br` command-line tool. Before backing up and restoring data, you need to [install the `br` command-line tool](/br/br-use-overview.md#deploy-and-use-br-command-line-tool) first.

> Snapshot backup is an implementation to back up the entire cluster. It is based on [multi-version concurrency control (MVCC)](/tidb-storage.md#mvcc) and backs up all data in the specified snapshot to a target storage. The size of the backup data is approximately the size of the compressed single replica in the cluster. After the backup is completed, you can restore the backup data to an empty cluster or a cluster that does not raise data conflicts (with the same schema or same tables), restore the cluster to the state at the time of the snapshot backup, and restore multiple replicas according to the cluster replica settings.

Reviewer: ```suggestion
Snapshot backup is an implementation to back up the entire cluster. It is based on [multi-version concurrency control (MVCC)](/tidb-storage.md#mvcc) and backs up all data in the specified snapshot to a target storage. The size of the backup data is approximately the size of the compressed single replica in the cluster. After the backup is completed, you can restore the backup data to an empty cluster or a cluster that does not contain conflict data (with the same schema or same tables), restore the cluster to the state at the time of the snapshot backup, and restore multiple replicas according to the cluster replica settings.
```

> Snapshot backup is an implementation to back up the entire cluster. It is based on [multi-version concurrency control (MVCC)](/tidb-storage.md#mvcc) and backs up all data in the specified snapshot to a target storage. The size of the backup data is approximately the size of the compressed single replica in the cluster. After the backup is completed, you can restore the backup data to an empty cluster or a cluster that does not raise data conflicts (with the same schema or same tables), restore the cluster to the state at the time of the snapshot backup, and restore multiple replicas according to the cluster replica settings.

Reviewer: ```suggestion
Snapshot backup is an implementation to back up the entire cluster. It is based on [multi-version concurrency control (MVCC)](/tidb-storage.md#mvcc) and backs up all data in the specified snapshot to a target storage. The size of the backup data is approximately the size of the compressed single replica in the cluster. After the backup is completed, you can restore the backup data to an empty cluster or a cluster that does not contain conflict data (with the same schema or same tables), restore the cluster to the time point of the snapshot backup, and restore multiple replicas according to the cluster replica settings.
```

Besides basic backup and restore, snapshot backup and restore also provides the following features:

* Backup data of a specified time point
* [Restore data of a specified database or table](#restore-a-database-or-a-table)

## Back up TiDB cluster snapshots

You can back up a TiDB cluster snapshot by running the `br backup full` command. Run `br backup full --help` to see the help information:

```shell
tiup br backup full --pd "${PD_IP}:2379" \
    --backupts '2022-09-08 13:30:00' \
    --storage "s3://backup-101/snapshot-202209081330?access_key=${access_key}&secret_access_key=${secret_access_key}" \
    --ratelimit 128 \
```

In the preceding command:

> - `--backupts`: The physical time of the snapshot. The format can be [TSO](/glossary.md#tso) or timestamp, such as `400036290571534337` or `2018-05-11 01:42:23`. If the data of this snapshot is garbage collected, the `br backup` command returns an error and `br` exits. If you leave this parameter unspecified, `br` picks the snapshot corresponding to the backup start time.

Reviewer: ```suggestion
- `--backupts`: The time point of the snapshot. The format can be [TSO](/glossary.md#tso) or timestamp, such as `400036290571534337` or `2018-05-11 01:42:23`. If the data of this snapshot is garbage collected, the `br backup` command returns an error and `br` exits. If you leave this parameter unspecified, `br` picks the snapshot corresponding to the backup start time.
```

- `--storage`: The storage address of the backup data. Snapshot backup supports Amazon S3, Google Cloud Storage, and Azure Blob Storage as backup storage. The preceding command uses Amazon S3 as an example. For more details, see [URL format of backup storages](/br/backup-and-restore-storages.md#url-format).
- `--ratelimit`: The maximum speed **per TiKV** performing backup tasks. The unit is in MiB/s.

During backup, a progress bar is displayed in the terminal as shown below. When the progress bar advances to 100%, the backup task is completed and statistics such as total backup time, average backup speed, and backup data size are displayed.

```shell
Full Backup <-------------------------------------------------------------------------------> 100.00%
Checksum <----------------------------------------------------------------------------------> 100.00%
*** ["Full Backup success summary"] *** [backup-checksum=3.597416ms] [backup-fast-checksum=2.36975ms] *** [total-take=4.715509333s] [BackupTS=435844546560000000] [total-kv=1131] [total-kv-size=250kB] [average-speed=53.02kB/s] [backup-data-size(after-compressed)=71.33kB] [Size=71330]
```

> ## Get the backup timestamp of a snapshot backup

Reviewer: ```suggestion
## Get the backup time point of a snapshot backup
```

> To manage the number of backups, if you need to get the physical time of a snapshot backup, you can run the following command:

Reviewer: ```suggestion
To manage lots of backups, if you need to get the physical time of a snapshot backup, you can run the following command:
```

```

### Restore tables in the `mysql` schema

Starting from `br` v5.1.0, when you back up snapshots, `br` backs up the **system tables** in the `mysql` schema and does not restore them by default. Starting from `br` v6.2.0, `br` restores **data in some system tables** if you configure `--with-sys-table`.

**The `br` command-line tool can restore data in the following system tables:**

```
+----------------------------------+
| mysql.columns_priv               |
| mysql.db                         |
| mysql.default_roles              |
| mysql.global_grants              |
| mysql.global_priv                |
| mysql.role_edges                 |
| mysql.tables_priv                |
| mysql.user                       |
+----------------------------------+
```

**The `br` command-line tool does not restore the following system tables:**

- Statistics tables (`mysql.stats_*`)
- System variable tables (`mysql.tidb` and `mysql.global_variables`)
- [Other system tables](https://github.com/pingcap/tidb/blob/master/br/pkg/restore/systable_restore.go#L31)

When you restore data related to system privilege, note the following:

> - `br` does not restore user data with `user` as `cloud_admin` and `host` as `'%'`. This user is reserved for TiDB Cloud. Do not create a user or role named `cloud_admin` in your cluster, because the user privileges related to `cloud_admin` cannot be restored correctly.

Reviewer: ```suggestion
- `br` does not restore user data with `user` as `cloud_admin` and `host` as `'%'`. This user is reserved for TiDB Cloud. Stop creating a user or role named `cloud_admin` in your cluster, because the user privileges related to `cloud_admin` cannot be restored correctly.
```

    - The target cluster has the same system tables as the backup data.
    - The **number of columns** in the system privilege table of the target cluster is the same as that in the backup data. The column order is not important.
    - The columns in the system privilege table of the target cluster are compatible with that in the backup data. If the data type of the column is a type with a length (such as integer and string), the length in the target cluster must be >= the length in the backup data. If the data type of the column is an `ENUM` type, the number of `ENUM` values in the target cluster must be a superset of that in the backup data.

## Performance and impact

### Performance and impact of snapshot backup

The backup feature has some impact on cluster performance (transaction latency and QPS). However, you can mitigate the impact by adjusting the number of backup threads [`backup.num-threads`](/tikv-configuration-file.md#num-threads-1) or by adding more clusters.

To illustrate the impact of backup, this document lists the test conclusions of several snapshot backup tests:

> - (5.3.0 and earlier) When the backup threads of `br` on a TiKV node take up 75% of the total CPU of the node, the QPS is reduced by 30% of the original QPS.

Reviewer 2: ```suggestion
- (5.3.0 and earlier) When the backup threads of `br` on a TiKV node take up 75% of the total CPU of the node, the QPS is reduced by 35% of the original QPS.
```

- Use the `--ratelimit` parameter to limit the speed of backup tasks. Note that this parameter limits the speed of **saving backup files to external storage**. When calculating the total size of backup files, use the `backup data size(after compressed)` as a benchmark.
- Adjust the TiKV configuration item [`backup.num-threads`](/tikv-configuration-file.md#num-threads-1) to limit the number of threads used by backup tasks. According to internal tests, when `br` uses no more than `8` threads for backup tasks, and the total CPU utilization of the cluster does not exceed 60%, the backup tasks have little impact on the cluster, regardless of the read and write workload.

The impact of backup on cluster performance can be reduced by limiting the backup threads number, but this affects the backup performance. The preceding tests show that the backup speed is proportional to the number of backup threads. When the number of threads is small, the backup speed is about 20 MiB/thread. For example, 5 backup threads on a single TiKV node can reach a backup speed of 100 MiB/s.

### Performance and impact of snapshot restore

> - During restoring data, TiDB tries to fully utilize the TiKV CPU, disk IO, and network bandwidth resources. Therefore, it is recommended to restore the backup data on an empty cluster to avoid affecting the running applications.

Reviewer 2: ```suggestion
- During data restore, TiDB tries to fully utilize the TiKV CPU, disk IO, and network bandwidth resources. Therefore, it is recommended to restore the backup data on an empty cluster to avoid affecting the running applications.
```

> - The restore speed of backup data has a great relationship with the cluster configuration, deployment, and running applications. In the internal tests, the restore speed of a single TiKV node can reach 100 MiB/s. The performance and impact of snapshot restore are different in different user scenarios and should be tested in actual environments.

Reviewer 2: ```suggestion
- The speed of restoring backup data is much related with the cluster configuration, deployment, and running applications. In internal tests, the restore speed of a single TiKV node can reach 100 MiB/s. The performance and impact of snapshot restore are varied in different user scenarios and should be tested in actual environments.
```

> ## What's next

Reviewer 2: ```suggestion
## See also
```

> * [TiDB back up and restore use cases](/br/backup-and-restore-use-cases.md)
> * [`br` command-line manual](/br/use-br-command-line-tool.md)
> * [TiDB snapshot backup and restore architecture](/br/br-snapshot-architecture.md)

Reviewer 2: ```suggestion
* [TiDB Backup and Restore Use Cases](/br/backup-and-restore-use-cases.md)
* [`br` Command-line Manual](/br/use-br-command-line-tool.md)
* [TiDB Snapshot Backup and Restore Architecture](/br/br-snapshot-architecture.md)
```
